subprojects/core/src/main/java/org/gradle/caching/internal/tasks/TarTaskOutputPacker.java

```diff
@@ -312,15 +328,23 @@ private void unpackPropertyEntry(ResolvedTaskOutputFilePropertySpec propertySpec
             outputFile = new File(propertyRoot, childPath);
         }
 
+        String internedPath = stringInterner.intern(outputFile.getAbsolutePath());
+        RelativePath relativePath = root ? RelativePath.EMPTY_ROOT : RelativePath.parse(!isDirEntry, childPath);
         if (isDirEntry) {
             FileUtils.forceMkdir(outputFile);
+            if (!root) {
+                fileSnapshots.put(propertyName, new DirectoryFileSnapshot(internedPath, relativePath, false));
```

Comment:
❌ The root directory will be detected as an overlapping output in the next run if we don't add it here. We also need to add a test that this works - probably by checking that there is no overlapping output message when we pulled the task from the cache and then run an up-to-date build again.